include/GLFW/glfw3.hpp

```diff
@@ -0,0 +1,177 @@
+
+#include <GLFW/glfw3.h>
+
+#include <functional>
+#include <optional>
+#include <string>
+#include <system_error>
+#include <vector>
+
+namespace GLFW_HPP_NAMESPACE {
+
+    namespace _private {
+#ifdef GLFW_HPP_NO_STD_STRING
+        inline GLFW_HPP_STRING makeString_(char const *str) { return str; }
+        inline char const *cString_(GLFW_HPP_STRING const &str) { return str; }
+#else
+        inline GLFW_HPP_STRING makeString_(char const *str) { return std::string(str); }
+        inline char const *cString_(GLFW_HPP_STRING const &str) { return str.c_str(); }
+#endif
+    }
+
+    using ErrorFun = GLFWerrorfun;
+
+    struct Size {
+        int width;
+        int height;
+    };
+
+    enum class ErrorCode {
+        noError = GLFW_NO_ERROR,
+        notInitialized = GLFW_NOT_INITIALIZED,
+        noCurrentContext = GLFW_NO_CURRENT_CONTEXT,
+        invalidEnum = GLFW_INVALID_ENUM,
+        invalidValue = GLFW_INVALID_VALUE,
+        outOfMemory = GLFW_OUT_OF_MEMORY,
+        apiUnavailable = GLFW_API_UNAVAILABLE,
+        versionUnavailable = GLFW_VERSION_UNAVAILABLE,
+        platformError = GLFW_PLATFORM_ERROR,
+        formatUnavailable = GLFW_FORMAT_UNAVAILABLE,
+        noWindowContext = GLFW_NO_WINDOW_CONTEXT,
+        cursorUnavailable = GLFW_CURSOR_UNAVAILABLE,
+        featureUnavailable = GLFW_FEATURE_UNAVAILABLE,
+        featureUnimplemented = GLFW_FEATURE_UNIMPLEMENTED,
+    };
+
+}
+
+namespace std {
+    template <>
+    struct is_error_code_enum<GLFW_HPP_NAMESPACE::ErrorCode> : public true_type {};
+}
+
+namespace GLFW_HPP_NAMESPACE {
+
+    class ErrorCategory : public std::error_category {
+    public:
+        const char *name() const _NOEXCEPT override {
+            return nullptr;
+        }
+
+        std::string message(int __ev) const override {
+            return std::string();
+        }
+    };
+
+    inline std::error_code make_error_code(ErrorCode code) {
+        return {static_cast<int>(code), ErrorCategory()};
+    }
+
+    class Error : public std::system_error {
+    public:
+        ErrorCode const code;
+
+    private:
+        Error(ErrorCode const code, GLFW_HPP_STRING const &description) :
+                std::system_error(std::error_code{code}, description),
+                code{code} {}
+
+        static std::optional<Error> get_() {
+            char const *description;
+            auto code = static_cast<ErrorCode const>(glfwGetError(&description));
+
+            if (code == ErrorCode::noError) {
+                return std::optional<Error>{};
+            }
+
+            return Error(code, description ? _private::makeString_(description) : "");
+        }
+
+        static void getAndThrow_() {
+            auto const err = get_();
+            if (err) {
+                throw err.value();
+            }
+        }
+
+        friend class GLFW;
+        friend class Window;
+    };
+
+    class Window {
+    public:
+        Window(Size const &size, GLFW_HPP_STRING const &title) :
```

Comment:
Here the lifetime of a `Window` object is tied the whether the window is closed or not, but the window is leaked. This does not follow the rule of 5, as you initialize a resource in the constructor.